storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,665 @@
+    return new SMAPredicate(
+        static_cast<const ScalarAttribute&>(predicate.getLeftOperand()).getAttribute().getID(),
+        predicate.getComparison().getComparisonID(),
+        predicate.getRightOperand().getStaticValue().makeReferenceToThis());
+  }
+}
+
+/**
+ * @brief Summation will promote lower precision types to higher precision types.
+ *
+ * @param type The type id of the type we wish to sum over.
+ * @return The higher precision sum typeid. Returns kNullType to indicate that the
+ *         given type cannot be summed (in lieu of a better sigil).
+ */
+inline TypeID sumType(TypeID type) {
+  switch (type) {
+    case kInt:
+    case kLong:
+      return kLong;
+    case kFloat:
+    case kDouble:
+      return kDouble;
+    default: {
+      return kNullType;
+    }
+  }
+}
+
+/**
+ * @param type A type id.
+ * @return True if the type can be summed.
+ */
+inline bool canSum(TypeID type) {
+  return sumType(type) != kNullType;
+}
+
+/**
+ * @brief Sets the given entry's sum to zero using the correct promoted type
+ *        of the sum. Does nothing if the entry's type cannot be summed.
+ * @note We ignore clearing any old data held in the previous sum because Double
+ *       and Long types are always represented inline.
+ *       
+ * @param entry A pointer to the entry to modify.
+ */
+inline void setTypedValueForSum(SMAEntry *entry) {
+  TypeID sum_type = sumType(entry->type_id);
+  if (sum_type == kLong) {
+    new (&entry->sum_aggregate) TypedValue(static_cast<std::int64_t>(0));
+  } else if (sum_type == kDouble) {
+    new (&entry->sum_aggregate) TypedValue(static_cast<double>(0.0));
+  }
+}
+
+/**
+ * @brief Correctly zeroes the SMA entry for each possible type.
+ *
+ * @param entry The entry for which the min/max references will be zeroed out.
+ */
+void setTypedValueForMinMax(SMAEntry *entry) {
+  TypedValue *min = &(entry->min_entry_ref.value);
+  TypedValue *max = &(entry->max_entry_ref.value);
+  switch (entry->type_id) {
+    case kInt:
+      new (min) TypedValue(static_cast<int>(0));
+      new (max) TypedValue(static_cast<int>(0));
+      return;
+    case kLong:
+      new (min) TypedValue(static_cast<std::int64_t>(0));
+      new (min) TypedValue(static_cast<std::int64_t>(0));
+      return;
+    case kFloat:
+      new (min) TypedValue(static_cast<float>(0));
+      new (max) TypedValue(static_cast<float>(0));
+    case kDouble:
+      new (min) TypedValue(static_cast<double>(0.0));
+      new (max) TypedValue(static_cast<double>(0.0));
+      return;
+    case kDatetimeInterval: {
+      DatetimeIntervalLit zero;
+      zero.interval_ticks = 0;
+      new (min) TypedValue(zero);
+      new (max) TypedValue(zero);
+      return;
+    }
+    case kYearMonthInterval: {
+      YearMonthIntervalLit zero;
+      zero.months = 0;
+      new (min) TypedValue(zero);
+      new (max) TypedValue(zero);
+      return;
+    }
+    default: {
+      CHECK(false) << "SMA Index was created on an unindexable type.";
```

Comment:
It may help to print the type ID along with the error message for better debugging.